File: Sources/TestingInternals/ExceptionHandling.cpp
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2024 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
//

#include "ExceptionHandling.h"

#if !SWT_NO_EXCEPTIONS
#include <cxxabi.h>
#include <exception>
#include <stdexcept>
#include <string.h>
#include <typeinfo>
#include <stdio.h>

void swt_cxxOnly_withExceptionHandling(
  void *context,
  void (* body)(void *context),
  void (* exceptionHandler)(void *context, void *exceptionPointer)
) {
  try {
    (* body)(context);
  } catch (...) {
    auto ep = std::current_exception();
    (* exceptionHandler)(context, &ep);
  }
}

char *swt_copyNameOfExceptionPointer(void *ep) {
  try {
    std::rethrow_exception(*reinterpret_cast<std::exception_ptr *>(ep));
  } catch (...) {
#if defined(__APPLE__) || defined(__linux__)
    if (auto type = __cxxabiv1::__cxa_current_exception_type()) {
      auto mangledName = type->name();
      int status = 0;
      if (auto demangledName = abi::__cxa_demangle(mangledName, nullptr, 0, &status)) {
        return demangledName;
      }
      return strdup(mangledName);
    }
#endif
  }

  return nullptr;
}

// MARK: - Test support

void swt_throwNumber(int value) {
  throw value;
}

void swt_throwCxxException(const char *what) {
  throw std::range_error(what);
}
#endif

File: Sources/Testing/Expectations/ExceptionHandling.swift
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2024 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
//

private import TestingInternals

@_spi(ForFoundationAndCxxStdlibOnly)
public enum CaughtException {
#if _runtime(_ObjC)
  case objectiveC(AnyObject)
#endif
  case cxx(UnsafeMutableRawPointer)
}

@available(*, unavailable)
extension CaughtException: Sendable {}

@_spi(ForFoundationAndCxxStdlibOnly)
public func withExceptionHandling(_ body: @escaping () -> Void, exceptionHandler: @escaping (borrowing CaughtException) -> Void) {
  struct Context {
    var body: () -> Void
    var exceptionHandler: (borrowing CaughtException) -> Void
  }
  var context = Context(body: body, exceptionHandler: exceptionHandler)
  withUnsafeMutablePointer(to: &context) { context in
#if _runtime(_ObjC)
    swt_withExceptionHandling(context) { context in
      let context = context!.load(as: Context.self)
      context.body()
    } objectiveCExceptionHandler: { context, exceptionPtr in
      let context = context!.load(as: Context.self)
      let exception = Unmanaged<AnyObject>.fromOpaque(exceptionPtr).takeUnretainedValue()
      context.exceptionHandler(.objectiveC(exception))
    } exceptionHandler: { context, exceptionPtr in
      let context = context!.load(as: Context.self)
      context.exceptionHandler(.cxx(exceptionPtr))
    }
#else
    swt_withExceptionHandling(context) { context in
      let context = context!.load(as: Context.self)
      context.body()
    } exceptionHandler: { context, exceptionPtr in
      let context = context!.load(as: Context.self)
      context.exceptionHandler(.cxx(exceptionPtr))
    }
#endif
  }
}

@_spi(ForFoundationAndCxxStdlibOnly)
public func name(ofExceptionPointer ep: UnsafeMutableRawPointer) -> String? {
  let typeName = swt_copyName(ofExceptionPointer: ep)
  defer {
    free(typeName)
  }
  return typeName.flatMap { String(validatingUTF8: $0) }
}
